File: flutter/sherpa_onnx/lib/src/offline_speaker_diarization.dart
// Copyright (c)  2024  Xiaomi Corporation
import 'dart:ffi';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import './sherpa_onnx_bindings.dart';
import './speaker_identification.dart';
import './offline_speaker_diarization_config.dart';

export './offline_speaker_diarization_config.dart';

/// Offline speaker diarizer.
class OfflineSpeakerDiarization {
  OfflineSpeakerDiarization.fromPtr(
      {required this.ptr, required this.config, required this.sampleRate});

  OfflineSpeakerDiarization._(
      {required this.ptr, required this.config, required this.sampleRate});

  /// Release the native diarizer.
  void free() {
    if (SherpaOnnxBindings.sherpaOnnxDestroyOfflineSpeakerDiarization == null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      return;
    }
    SherpaOnnxBindings.sherpaOnnxDestroyOfflineSpeakerDiarization?.call(ptr);
    ptr = nullptr;
  }

  /// Create a diarizer from [config].
  factory OfflineSpeakerDiarization(OfflineSpeakerDiarizationConfig config) {
    if (SherpaOnnxBindings.sherpaOnnxCreateOfflineSpeakerDiarization == null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    final c = calloc<SherpaOnnxOfflineSpeakerDiarizationConfig>();

    c.ref.segmentation.pyannote.model =
        config.segmentation.pyannote.model.toNativeUtf8();
    c.ref.segmentation.numThreads = config.segmentation.numThreads;
    c.ref.segmentation.debug = config.segmentation.debug ? 1 : 0;
    c.ref.segmentation.provider = config.segmentation.provider.toNativeUtf8();

    c.ref.embedding.model = config.embedding.model.toNativeUtf8();
    c.ref.embedding.numThreads = config.embedding.numThreads;
    c.ref.embedding.debug = config.embedding.debug ? 1 : 0;
    c.ref.embedding.provider = config.embedding.provider.toNativeUtf8();

    c.ref.clustering.numClusters = config.clustering.numClusters;
    c.ref.clustering.threshold = config.clustering.threshold;

    c.ref.minDurationOn = config.minDurationOn;
    c.ref.minDurationOff = config.minDurationOff;

    final ptr =
        SherpaOnnxBindings.sherpaOnnxCreateOfflineSpeakerDiarization?.call(c) ??
            nullptr;

    calloc.free(c.ref.embedding.provider);
    calloc.free(c.ref.embedding.model);
    calloc.free(c.ref.segmentation.provider);
    calloc.free(c.ref.segmentation.pyannote.model);
    calloc.free(c);

    if (ptr == nullptr) {
      throw Exception(
          "Failed to create offline speaker diarization. Please check your config");
    }

    int sampleRate = SherpaOnnxBindings
              .sherpaOnnxOfflineSpeakerDiarizationGetSampleRate
              ?.call(ptr) ?? 0;

    return OfflineSpeakerDiarization._(
        ptr: ptr, config: config, sampleRate: sampleRate);
  }

  /// Process a complete waveform and return speaker-labeled segments.
  List<OfflineSpeakerDiarizationSegment> process(
      {required Float32List samples}) {
    if (SherpaOnnxBindings.sherpaOnnxOfflineSpeakerDiarizationProcess == null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      return <OfflineSpeakerDiarizationSegment>[];
    }

    final n = samples.length;
    final Pointer<Float> p = calloc<Float>(n);
    try {
      final pList = p.asTypedList(n);
      pList.setAll(0, samples);

      final r = SherpaOnnxBindings.sherpaOnnxOfflineSpeakerDiarizationProcess
              ?.call(ptr, p, n) ??
          nullptr;
      try {
        return _processImpl(r);
      } finally {
        if (r != nullptr) {
          SherpaOnnxBindings.sherpaOnnxOfflineSpeakerDiarizationDestroyResult
              ?.call(r);
        }
      }
    } finally {
      calloc.free(p);
    }
  }

  List<OfflineSpeakerDiarizationSegment> processWithCallback({
    required Float32List samples,
    required int Function(int numProcessedChunks, int numTotalChunks) callback,
  }) {
    if (SherpaOnnxBindings
            .sherpaOnnxOfflineSpeakerDiarizationProcessWithCallbackNoArg ==
        null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      return <OfflineSpeakerDiarizationSegment>[];
    }

    final n = samples.length;
    final Pointer<Float> p = calloc<Float>(n);
    try {
      final pList = p.asTypedList(n);
      pList.setAll(0, samples);

      final wrapper = NativeCallable<
              SherpaOnnxOfflineSpeakerDiarizationProgressCallbackNoArgNative>.isolateLocal(
          (int numProcessedChunks, int numTotalChunks) {
        return callback(numProcessedChunks, numTotalChunks);
      }, exceptionalReturn: 0);
      try {
        final r = SherpaOnnxBindings
                .sherpaOnnxOfflineSpeakerDiarizationProcessWithCallbackNoArg
                ?.call(ptr, p, n, wrapper.nativeFunction) ??
            nullptr;
        try {
          return _processImpl(r);
        } finally {
          if (r != nullptr) {
            SherpaOnnxBindings.sherpaOnnxOfflineSpeakerDiarizationDestroyResult
                ?.call(r);
          }
        }
      } finally {
        wrapper.close();
      }
    } finally {
      calloc.free(p);
    }
  }

  List<OfflineSpeakerDiarizationSegment> _processImpl(
      Pointer<SherpaOnnxOfflineSpeakerDiarizationResult> r) {
    if (r == nullptr) {
      return <OfflineSpeakerDiarizationSegment>[];
    }

    final numSegments = SherpaOnnxBindings
            .sherpaOnnxOfflineSpeakerDiarizationResultGetNumSegments
            ?.call(r) ??
        0;
    final segments = SherpaOnnxBindings
            .sherpaOnnxOfflineSpeakerDiarizationResultSortByStartTime
            ?.call(r) ??
        nullptr;

    if (segments == nullptr) {
      return <OfflineSpeakerDiarizationSegment>[];
    }

    final ans = <OfflineSpeakerDiarizationSegment>[];
    for (int i = 0; i != numSegments; ++i) {
      final s = segments + i;

      final tmp = OfflineSpeakerDiarizationSegment(
          start: s.ref.start, end: s.ref.end, speaker: s.ref.speaker);
      ans.add(tmp);
    }

    SherpaOnnxBindings.sherpaOnnxOfflineSpeakerDiarizationDestroySegment
        ?.call(segments);

    return ans;
  }

  Pointer<SherpaOnnxOfflineSpeakerDiarization> ptr;
  OfflineSpeakerDiarizationConfig config;
  final int sampleRate;
}

File: sherpa-onnx/rust/sherpa-onnx/src/offline_speaker_diarization.rs
//! Offline speaker diarization.
//!
//! This combines segmentation, speaker embedding extraction, and clustering.
//! See `rust-api-examples/examples/offline_speaker_diarization.rs`.

use crate::{speaker_embedding::SpeakerEmbeddingExtractorConfig, utils::to_c_ptr};
use sherpa_onnx_sys as sys;
use std::ffi::CString;
use std::slice;

#[derive(Clone, Debug, Default)]
/// Pyannote segmentation model path.
pub struct OfflineSpeakerSegmentationPyannoteModelConfig {
    pub model: Option<String>,
}

impl OfflineSpeakerSegmentationPyannoteModelConfig {
    fn to_sys(
        &self,
        cstrings: &mut Vec<CString>,
    ) -> sys::OfflineSpeakerSegmentationPyannoteModelConfig {
        sys::OfflineSpeakerSegmentationPyannoteModelConfig {
            model: to_c_ptr(&self.model, cstrings),
        }
    }
}

#[derive(Clone, Debug)]
/// Segmentation model configuration for diarization.
pub struct OfflineSpeakerSegmentationModelConfig {
    pub pyannote: OfflineSpeakerSegmentationPyannoteModelConfig,
    pub num_threads: i32,
    pub debug: bool,
    pub provider: Option<String>,
}

impl Default for OfflineSpeakerSegmentationModelConfig {
    fn default() -> Self {
        Self {
            pyannote: Default::default(),
            num_threads: 1,
            debug: false,
            provider: Some("cpu".to_string()),
        }
    }
}

impl OfflineSpeakerSegmentationModelConfig {
    fn to_sys(&self, cstrings: &mut Vec<CString>) -> sys::OfflineSpeakerSegmentationModelConfig {
        sys::OfflineSpeakerSegmentationModelConfig {
            pyannote: self
                .pyannote
                .to_sys(cstrings),
            num_threads: self.num_threads,
            debug: self.debug as i32,
            provider: to_c_ptr(&self.provider, cstrings),
        }
    }
}

#[derive(Clone, Debug)]
/// Fast clustering options used after segmentation and embedding extraction.
pub struct FastClusteringConfig {
    pub num_clusters: i32,
    pub threshold: f32,
}

impl Default for FastClusteringConfig {
    fn default() -> Self {
        Self {
            num_clusters: -1,
            threshold: 0.5,
        }
    }
}

impl FastClusteringConfig {
    fn to_sys(&self) -> sys::FastClusteringConfig {
        sys::FastClusteringConfig {
            num_clusters: self.num_clusters,
            threshold: self.threshold,
        }
    }
}

#[derive(Clone, Debug)]
/// Top-level configuration for [`OfflineSpeakerDiarization`].
pub struct OfflineSpeakerDiarizationConfig {
    pub segmentation: OfflineSpeakerSegmentationModelConfig,
    pub embedding: SpeakerEmbeddingExtractorConfig,
    pub clustering: FastClusteringConfig,
    pub min_duration_on: f32,
    pub min_duration_off: f32,
}

impl Default for OfflineSpeakerDiarizationConfig {
    fn default() -> Self {
        Self {
            segmentation: Default::default(),
            embedding: Default::default(),
            clustering: Default::default(),
            min_duration_on: 0.3,
            min_duration_off: 0.5,
        }
    }
}

impl OfflineSpeakerDiarizationConfig {
    fn to_sys(&self, cstrings: &mut Vec<CString>) -> sys::OfflineSpeakerDiarizationConfig {
        sys::OfflineSpeakerDiarizationConfig {
            segmentation: self
                .segmentation
                .to_sys(cstrings),
            embedding: self
                .embedding
                .to_sys(cstrings),
            clustering: self
                .clustering
                .to_sys(),
            min_duration_on: self.min_duration_on,
            min_duration_off: self.min_duration_off,
        }
    }
}

#[derive(Clone, Debug)]
/// One diarization segment labeled with a speaker index.
pub struct OfflineSpeakerDiarizationSegment {
    pub start: f32,
    pub end: f32,
    pub speaker: i32,
}

/// Offline speaker diarizer.
pub struct OfflineSpeakerDiarization {
    ptr: *const sys::OfflineSpeakerDiarization,
}

// SAFETY: The sherpa-onnx C library is thread-safe for single-object usage.
unsafe impl Send for OfflineSpeakerDiarization {}
unsafe impl Sync for OfflineSpeakerDiarization {}

impl OfflineSpeakerDiarization {
    /// Create a diarizer from `config`.
    pub fn create(config: &OfflineSpeakerDiarizationConfig) -> Option<Self> {
        let mut cstrings = Vec::new();
        let sys_config = config.to_sys(&mut cstrings);
        let ptr = unsafe { sys::SherpaOnnxCreateOfflineSpeakerDiarization(&sys_config) };
        if ptr.is_null() {
            None
        } else {
            Some(Self { ptr })
        }
    }

    /// Return the sample rate expected by the segmentation model.
    pub fn sample_rate(&self) -> i32 {
        unsafe { sys::SherpaOnnxOfflineSpeakerDiarizationGetSampleRate(self.ptr) }
    }

    /// Replace the current configuration.
    pub fn set_config(&self, config: &OfflineSpeakerDiarizationConfig) {
        let mut cstrings = Vec::new();
        let sys_config = config.to_sys(&mut cstrings);
        unsafe { sys::SherpaOnnxOfflineSpeakerDiarizationSetConfig(self.ptr, &sys_config) }
    }

    /// Process a complete waveform and return a diarization result.
    pub fn process(&self, samples: &[f32]) -> Option<OfflineSpeakerDiarizationResult> {
        let ptr = unsafe {
            sys::SherpaOnnxOfflineSpeakerDiarizationProcess(
                self.ptr,
                samples.as_ptr(),
                samples.len() as i32,
            )
        };
        if ptr.is_null() {
            None
        } else {
            Some(OfflineSpeakerDiarizationResult { ptr })
        }
    }
}

impl Drop for OfflineSpeakerDiarization {
    fn drop(&mut self) {
        unsafe {
            if !self
                .ptr
                .is_null()
            {
                sys::SherpaOnnxDestroyOfflineSpeakerDiarization(self.ptr);
            }
        }
    }
}

/// Result object returned by [`OfflineSpeakerDiarization::process`].
pub struct OfflineSpeakerDiarizationResult {
    ptr: *const sys::OfflineSpeakerDiarizationResult,
}

// SAFETY: The sherpa-onnx C library is thread-safe for single-object usage.
unsafe impl Send for OfflineSpeakerDiarizationResult {}
unsafe impl Sync for OfflineSpeakerDiarizationResult {}

impl OfflineSpeakerDiarizationResult {
    /// Return the number of speakers estimated for the recording.
    pub fn num_speakers(&self) -> i32 {
        unsafe { sys::SherpaOnnxOfflineSpeakerDiarizationResultGetNumSpeakers(self.ptr) }
    }

    /// Return the number of diarization segments.
    pub fn num_segments(&self) -> i32 {
        unsafe { sys::SherpaOnnxOfflineSpeakerDiarizationResultGetNumSegments(self.ptr) }
    }

    /// Return all segments sorted by start time.
    pub fn sort_by_start_time(&self) -> Vec<OfflineSpeakerDiarizationSegment> {
        let n = self.num_segments();
        if n <= 0 {
            return Vec::new();
        }

        unsafe {
            let p = sys::SherpaOnnxOfflineSpeakerDiarizationResultSortByStartTime(self.ptr);
            if p.is_null() {
                return Vec::new();
            }

            let segments = slice::from_raw_parts(p, n as usize)
                .iter()
                .map(|s| OfflineSpeakerDiarizationSegment {
                    start: s.start,
                    end: s.end,
                    speaker: s.speaker,
                })
                .collect::<Vec<_>>();
            sys::SherpaOnnxOfflineSpeakerDiarizationDestroySegment(p);
            segments
        }
    }
}

impl Drop for OfflineSpeakerDiarizationResult {
    fn drop(&mut self) {
        unsafe {
            if !self
                .ptr
                .is_null()
            {
                sys::SherpaOnnxOfflineSpeakerDiarizationDestroyResult(self.ptr);
            }
        }
    }
}
